File: java/demo/src/main/java/com/yuanzhibang/demo/api/ApiRequestHelper.java
package com.yuanzhibang.demo.api;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.apache.http.HttpHost;
import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.auth.AuthScope;
import org.apache.http.auth.UsernamePasswordCredentials;
import org.apache.http.client.CredentialsProvider;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.BasicCredentialsProvider;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.client.ProxyAuthenticationStrategy;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;

import com.google.gson.Gson;

public class ApiRequestHelper {

    @SuppressWarnings("unchecked")
    public static Object post(String url, Map<String, String> params, Map<String, String> headers,
            boolean parseJson, RequestProxy proxy) {
        try {
            CloseableHttpClient httpClient;
            HttpPost request = new HttpPost(url);
            if (proxy != null) {
                CredentialsProvider credsProvider = new BasicCredentialsProvider();
                credsProvider.setCredentials(
                        new AuthScope(proxy.host, proxy.port),
                        new UsernamePasswordCredentials(proxy.username, proxy.password));
                httpClient = HttpClients.custom()
                        .setDefaultCredentialsProvider(credsProvider)
                        .setProxyAuthenticationStrategy(ProxyAuthenticationStrategy.INSTANCE)
                        .build();
                HttpHost proxyHost = new HttpHost(proxy.host, proxy.port);

                RequestConfig config = RequestConfig.custom()
                        .setProxy(proxyHost)
                        .build();
                request.setConfig(config);

            } else {
                httpClient = HttpClients.custom()
                        .build();
            }

            request.setHeader("Content-Type", "multipart/form-data");
            if (params != null) {
                List<NameValuePair> postParams = new ArrayList<>();
                for (Map.Entry<String, String> entry : params.entrySet()) {
                    postParams.add(new BasicNameValuePair(entry.getKey(), entry.getValue()));
                }
                request.setEntity(new UrlEncodedFormEntity(postParams));
            }
            if (headers != null) {
                for (Map.Entry<String, String> entry : headers.entrySet()) {
                    request.setHeader(entry.getKey(), entry.getValue());
                }
            }
            try {
                HttpResponse response = httpClient.execute(request);
                String responseString = EntityUtils.toString(response.getEntity());
                try {
                    if (parseJson) {
                        Gson gson = new Gson();
                        Map<String, String> info = gson.fromJson(responseString, Map.class);
                        return info;
                    } else {
                        return responseString;
                    }
                } catch (Exception e) {
                    return null;
                }
            } catch (IOException e) {
                return null;
            } finally {
                request.releaseConnection();
            }
        } catch (UnsupportedEncodingException e) {
            return null;
        }

    }

    // public static void test() {
    // RequestProxy proxy = new RequestProxy("demo-proxy", 7789, "123", "12345678");
    // ApiRequestHelper.post("https://api-service.yuanzhibang.com/api/v1/Ip/getClientIp",
    // null, null, proxy);
    // }
}

File: python/api_request_helper.py
import requests


class OauthError(Exception):

    def __init__(self, code, message, response):
        self.code = code
        self.message = message
        self.response = response


class OauthNetworkError(OauthError):
    pass


class OauthApiError(OauthError):
    pass


class ApiRequestHelper:
    @staticmethod
    def post(url, data, proxies=None):
        response = requests.post(url=url, data=data, proxies=proxies)
        ApiRequestHelper.raise_network_error(response)
        ApiRequestHelper.raise_api_error(response)
        response_object = response.json()
        return response_object['data']

    @staticmethod
    def raise_network_error(response):
        status_code = response.status_code
        if status_code != 200:
            raise OauthNetworkError(status_code, 'network error', response)

    @ staticmethod
    def raise_api_error(response):
        try:
            response_object = response.json()
            status: str = response_object['status']
            message = response_object['message']
        except Exception as e:
            raise OauthNetworkError("0000", 'network error', response)
        if status.startswith('2') is False:
            raise OauthApiError(status, message, response)

    @ staticmethod
    def test():
        ApiRequestHelper.post('https://api-service.yuanzhibang.com/api/v1/Ip/getClientIp', {},
                              {
            "http": '123:12345678@demo-proxy:7789',
            "https": '123:12345678@demo-proxy:7789'
        })
